### PR TITLE
Convert Sinope light to new style AttributeDefs, add ZHA events

### PR DESCRIPTION
**Pull request by ckm2k1**
## Proposed change
Add ability for Sinope light switches to handle routing attribute reports on the manufacturer cluster attribute `action_report` as ZHA events.

## Additional information
Related PR zigpy/zha/pull/153 made it so that `action_report` doesn't need to be manually configured for reporting (as shown by @claudegel here: https://github.com/claudegel/sinope-zha?tab=readme-ov-file#a-create-the-reporting-).

## Checklist

- [x] The changes are tested and work correctly
- [x] `pre-commit` checks pass / the code has been formatted using Black
- [x] Tests have been added to verify that the new code works

### CODE REVIEW

File: tests/test_sinope.py [hunk 1]
```diff
@@ -57,3 +72,135 @@ async def test_sinope_flow_measurement(zigpy_device_from_quirk, quirk):
         == flow_measurement_other_attr_id
     )
     assert flow_measurement_listener.attribute_updates[1][1] == 25  # not modified
+
+
+def _get_xbee_packet_data(
```
Comment (TheJulianJES) [inline tests/test_sinope.py]: Can we replace the "xbee" part here?

File: zhaquirks/sinope/light.py [hunk 2]
```diff
@@ -96,6 +110,67 @@ class Action(t.enum8):
...
+        )
+
+        if hdr.command_id != foundation.GeneralCommand.Report_Attributes:
+            return super().handle_cluster_general_request(
+                hdr, args, dst_addressing=dst_addressing
+            )
+
+        attr = args[0][0]
+
+        if attr.attrid != self.attributes_by_name["action_report"].id:
```
Comment (TheJulianJES) [inline zhaquirks/sinope/light.py]: Ideally, we'd use the new zigpy-style `AttributeDefs` for the attributes (and then reference them here.
Can be cleaned up in another PR though, as you're not adding new attributes in this one.

File: tests/test_sinope.py (hunk 1 above)
Comment (ckm2k1) [reply tests/test_sinope.py]: Yep, replaced.

File: zhaquirks/sinope/light.py (hunk 2 above)
Comment (ckm2k1) [reply zhaquirks/sinope/light.py]: SG, will do it in a follow up PR.

Comment (ckm2k1) [reply zhaquirks/sinope/light.py]: Ended up refactoring to use the new style `AttributeDefs` in this PR.

**Comment by claudegel** [comment]
So we will need to do the same in sinope thermostat.py, sensor.py and switch.py
Can I help to reformat some of them ?

**Comment by claudegel** [comment]
Ok I'm almost finish with the changes for thermostat.py, switch.py and sensor.py.
This is clear for the manufacturer cluster but in those files there are attribute addition in other cluster like Thermostat cluster
```
    attributes = Thermostat.attributes.copy()
    attributes.update(
        {
            0x0400: ("set_occupancy", Occupancy, True),
            0x0401: ("main_cycle_output", CycleOutput, True),
            ...
```
How do you change the AttributeDefs
```
    class AttributeDefs(foundation.BaseAttributeDefs):
        """Sinope Thermostat Custom Cluster Attributes."""

        set_occupancy: Final = foundation.ZCLAttributeDef(
            id=0x0400, type=Occupancy, access="rw", is_manufacturer_specific=True
        )
        ...
        )

        attributes = Thermostat.attributes.copy()
        attributes.update(
        {
        set_occupancy
        ...
        }
```

**Comment by ckm2k1** [comment]
@claudegel I don't know for sure, but my guess would be that you can probably inherit from the ZHA thermostat cluster attribute defs, something like this:

```python
class AttributeDefs(Thermostat.AttributeDefs):
 ...
```


Btw, thank you for all the work you've done on supporting Sinope devices in ZHA!

**Comment by claudegel** [comment]
Thanks I'll push a PR tomorrow for thermostat, switch and sensor

**Comment by claudegel** [comment]
I've tried this but the attribute is not added to IasZone cluster:
```
class SinopeTechnologiesIasZoneCluster(CustomCluster, IasZone):
    """SinopeTechnologiesIasZoneCluster custom cluster."""

    LeakStatus: Final = LeakStatus

    class AttributeDefs(IasZone.AttributeDefs):
        """Sinope Manufacturer IasZone Cluster Attributes."""

    leak_status: Final = foundation.ZCLAttributeDef(
        id=0x0030, type=LeakStatus, access="rw", is_manufacturer_specific=True
    )
```
Any idea ?

**Comment by claudegel** [comment]
Was able to fix like this: 
Attribute is added to IasZone cluster but I still have an error
```
    leak_status: Final = foundation.BaseAttributeDefs(
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: BaseAttributeDefs() takes no arguments
```
```
    class AttributeDefs(IasZone.AttributeDefs):
        """Sinope Manufacturer IasZone Cluster Attributes."""

    leak_status: Final = foundation.BaseAttributeDefs(
        id=0x0030, type=LeakStatus, access="rw", is_manufacturer_specific=True
    )
```
If I replace foundation.BaseAttributeDefs by foundation.ZCLAttributeDef it is not working and atribute leak_status is not added
 I've another strange thing. In the SinopeManufacturerCluster I have some Unknown_attr_1... attributes that I'm still working on . I need them to appear in the cluster but with this new style AttributeDefs they are just skipped. All other named attributes are there but the unknown one are missing. Is it because there is a number in them ?

**Comment by claudegel** [comment]
I got it ... a stupid indentation that was not noticed in the log
ready for a PR

**Comment by ckm2k1** [comment]
@TheJulianJES - Anything missing in this PR to merge up?

**Comment by claudegel** [comment]
Maybe change line
```
cluster_revision: Final = foundation.ZCLAttributeDef(
            id=0xFFFD, type=t.uint16_t, access="r", is_manufacturer_specific=True
        )
```
by
`    cluster_revision: Final = foundation.ZCL_CLUSTER_REVISION_ATTR`

**Comment by ckm2k1** [comment]
> Maybe change line
> 
> ```
> cluster_revision: Final = foundation.ZCLAttributeDef(
>             id=0xFFFD, type=t.uint16_t, access="r", is_manufacturer_specific=True
>         )
> ```
> 
> by ` cluster_revision: Final = foundation.ZCL_CLUSTER_REVISION_ATTR`

Good point, changed it.

**Comment by ckm2k1** [comment]
@TheJulianJES - another bump, anything else needed to merge this? 
Review feedback is addressed, all tests are passing.

**Comment by TheJulianJES** [comment]
There's no need to ping me. It'll be looked at when I (or someone else) has time.
The next zha-quirks release will only be at the end of the month anyway, as they're always made a couple of days before the next major HA beta is released.

**Comment by claudegel** [comment]
Look good. Thank you

File: tests/test_sinope.py [hunk 3]
```diff
@@ -57,3 +73,135 @@ async def test_sinope_flow_measurement(zigpy_device_from_quirk, quirk):
...
+    bind_patch = mock.patch("zigpy.zcl.Cluster.bind", mock.AsyncMock())
+
+    with request_patch as request_mock, bind_patch as bind_mock:
+        request_mock.return_value = (foundation.Status.SUCCESS, "done")
+
+        await manu_cluster.bind()
+        await manu_cluster.configure_reporting(
+            SinopeTechnologiesManufacturerCluster.attributes_by_name[
+                "action_report"
+            ].id,
```
Comment (TheJulianJES) [inline tests/test_sinope.py]: We can use `AttributeDefs` here now.
```suggestion
            SinopeTechnologiesManufacturerCluster.AttributeDefs.action_report.id
```

File: tests/test_sinope.py [hunk 4]
```diff
@@ -57,3 +73,135 @@ async def test_sinope_flow_measurement(zigpy_device_from_quirk, quirk):
...
+    )
+    data = _get_packet_data(foundation.GeneralCommand.Report_Attributes, attr)
+    device.handle_message(260, cluster_id, endpoint_id, endpoint_id, data)
+    # ZHA event emitted because we pass non "action_report"
+    # reports to the base class handler.
+    assert cluster_listener.zha_send_event.call_count == 1
+
+
+@pytest.mark.parametrize("quirk", (SinopeTechnologieslight,))
+async def test_sinope_light_switch_reporting(zigpy_device_from_quirk, quirk):
```
Comment (TheJulianJES) [inline tests/test_sinope.py:184]: Tbh, I'm not sure we really need this test in zha quirks? We're not really testing any quirk specific code here, right?
(maybe except that the attribute exists)

We can leave this in for now though.

File: zhaquirks/sinope/light.py [hunk 5]
```diff
@@ -23,79 +27,194 @@
...
+        self,
+        hdr: foundation.ZCLHeader,
+        args: list[Any],
+        *,
+        dst_addressing: Optional[
+            Union[t.Addressing.Group, t.Addressing.IEEE, t.Addressing.NWK]
+        ] = None,
+    ):
+        """Handle the cluster command."""
+        self.debug(
```
Comment (TheJulianJES) [inline zhaquirks/sinope/light.py:176]: We can leave it like this for now, but this causes a `zigpy` log. We might want to use our own logger here, since that'll clearly be `zhaquirks.sinope`. Other quirks aren't consistent about this though.
The string does include "Sinope" at least, so that's good.

**Review by TheJulianJES — COMMENTED** [review]
`AttributeDefs` changes look good. The `action_report` stuff seems fine as well, I think.

File: tests/test_sinope.py [hunk 6]
```diff
@@ -57,3 +73,133 @@ async def test_sinope_flow_measurement(zigpy_device_from_quirk, quirk):
...
+
+    request_patch = mock.patch("zigpy.zcl.Cluster.request", mock.AsyncMock())
+    bind_patch = mock.patch("zigpy.zcl.Cluster.bind", mock.AsyncMock())
+
+    with request_patch as request_mock, bind_patch as bind_mock:
+        request_mock.return_value = (foundation.Status.SUCCESS, "done")
+
+        await manu_cluster.bind()
+        await manu_cluster.configure_reporting(
+            SinopeTechnologiesManufacturerCluster.AttributeDefs.action_report.id
```
Comment (TheJulianJES) [inline tests/test_sinope.py]: Add back missing comma I removed:
```suggestion
            SinopeTechnologiesManufacturerCluster.AttributeDefs.action_report.id,
```

File: zhaquirks/sinope/__init__.py [hunk 7]
```diff
@@ -23,7 +23,7 @@
 
 SINOPE = "Sinope Technologies"
 SINOPE_MANUFACTURER_CLUSTER_ID = 0xFF01
-ATTRIBUTE_ACTION = "actionReport"
+ATTRIBUTE_ACTION = "action_report"
```
Comment (TheJulianJES) [inline zhaquirks/sinope/__init__.py:26]: Changing this might cause issues with current automations that do not use device triggers, right?

**Review by TheJulianJES — APPROVED** [review]
Thanks, let's try this.
